test/integration/replicaset/replicaset_test.go

```diff
@@ -272,246 +279,698 @@ func TestAdoption(t *testing.T) {
+		}
+		*newRS.Spec.Replicas = replicas
+		rs, err = rsClient.Update(newRS)
+		return err
+	}); err != nil {
+		t.Fatalf("Failed to update .Spec.Replicas to %d for rs %s: %v", replicas, rs.Name, err)
 	}
-	t.Logf("patched rs = %#v", rs)
-	// wait for the rs select both pods and delete one of them
-	if err := wait.Poll(10*time.Second, 60*time.Second, func() (bool, error) {
-		return verifyRemainingObjects(t, clientSet, ns.Name, 1, 1)
+	waitRSStable(t, c, rs)
+}
+
+func updatePod(t *testing.T, podClient typedv1.PodInterface, pod *v1.Pod, updateFunc func(*v1.Pod)) {
+	if err := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		newPod, err := podClient.Get(pod.Name, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+		updateFunc(newPod)
+		_, err = podClient.Update(newPod)
+		return err
 	}); err != nil {
-		t.Fatal(err)
+		t.Fatalf("Failed to update pod %s: %v", pod.Name, err)
 	}
-	close(stopCh)
 }
 
-func TestUpdateSelectorToRemoveControllerRef(t *testing.T) {
-	// We have pod1, pod2 and rs. rs.spec.replicas=2. At first rs.Selector
-	// matches pod1 and pod2; change the selector to match only pod1. Verify
-	// that rs creates one more pod, so there are 3 pods. Also verify that
-	// pod2's controllerRef is cleared.
-	s, closeFn, rm, informers, clientSet := rmSetup(t)
+func getPods(t *testing.T, podClient typedv1.PodInterface, labelMap map[string]string) *v1.PodList {
+	podSelector := labels.Set(labelMap).AsSelector()
+	options := metav1.ListOptions{LabelSelector: podSelector.String()}
+	pods, err := podClient.List(options)
+	if err != nil {
+		t.Fatalf("Failed obtaining a list of pods that match the pod labels %v: %v", labelMap, err)
+	}
+	if pods == nil {
+		t.Fatalf("Obtained a nil list of pods")
+	}
+	return pods
+}
+
+func TestReplicaSetBasics(t *testing.T) {
+	s, closeFn, rm, informers, c := rmSetup(t)
 	defer closeFn()
-	podInformer := informers.Core().V1().Pods().Informer()
-	ns := framework.CreateTestingNamespace("rs-update-selector-to-remove-controllerref", s, t)
+	ns := framework.CreateTestingNamespace("test-replicaset-basics", s, t)
 	defer framework.DeleteTestingNamespace(ns, s, t)
+	labelMap := map[string]string{"foo": "bar"}
 	rs := newRS("rs", ns.Name, 2)
-	pod1 := newMatchingPod("pod1", ns.Name)
-	pod1.Labels["uniqueKey"] = "1"
-	pod2 := newMatchingPod("pod2", ns.Name)
-	pod2.Labels["uniqueKey"] = "2"
-	createRSsPods(t, clientSet, []*v1beta1.ReplicaSet{rs}, []*v1.Pod{pod1, pod2}, ns.Name)
+	rs.Spec.Selector.MatchLabels = labelMap
+	rs.Spec.Template.Labels = labelMap
+	deletingPod := newMatchingPod("deleting-pod", ns.Name)
+	deletingPod.Labels = labelMap
+	deletingPod.DeletionTimestamp = &metav1.Time{Time: time.Now().Add(10 * time.Minute)}
+	deletingPod.Finalizers = []string{"foregroundDeletion"}
+	failedPod := newMatchingPod("failed-pod", ns.Name)
+	failedPod.Labels = labelMap
+	failedPod.Status.Phase = v1.PodFailed
+	rss, _ := createRSsPods(t, c, []*v1beta1.ReplicaSet{rs}, []*v1.Pod{deletingPod, failedPod})
```

Comment:
You're creating these Pods as orphans, so it only checks that RS will not adopt them.

The more common scenario that we want to test is when an existing Pod already owned by the RS goes into Failed phase or begins graceful deletion. To test that, it would be better to let the RS create the Pods as in other tests. Then you grab those Pods and set one to Failed (note you must use Pods().UpdateStatus() for this) and add the finalizer to another before trying to delete it.

In this scenario, RS should react by noticing the Pods change state and immediately creating new Pods to replace them.

Reply:
Yes, `UpdateStatus()` is the right call, not `Update`. It explains why the pod disappears after scaling (or no pod is created to replace the failed pod) - the pod does not have its status updated!

If we want to let RS creates the pods, it means `DeletionTimestamp` is not an option here. The reason I decide to create the deleting pod myself is because I want to ensure setting `DeletionTimestamp` does not break the validation (no validation error for explict creation of deletion pod). However, I agree that letting RS to create the pod itself is the correct way. In other words, `gracePeriodSeconds` of `DeleteOptions` + finalizer is the way to go.